lib/checks/aria/allowed-role.js

```diff
@@ -0,0 +1,180 @@
+ * @param  {string}  tagName key of roleSubset
+ * @param  {string}  role    current role
+ * @return {Boolean}         True if allowed
+ */
+function withinSubset(tagName, role) {
+	if(!roleSubset[tagName]) {
+		return false;
+	}
+
+	return roleSubset[tagName].indexOf(role) > -1;
+}
+
+/**
+ * Contains a function for each htmlTag where not all roles allowed
+ * @type {Object}
+ */
+var elementCheck = {
+	a: (node, role) => {
+		if(node.namespaceURI === 'http://www.w3.org/2000/svg') {
+			return true;
+		}
+
+		if(node.href) {
+			return withinSubset('aWithHref', role);
+		} else {
+			return true;
+		}
+	},
+	area: (node) => !(node.href),
+	button: (node, role) => {
+		if(node.getAttribute('type') === 'menu' ) {
+			return role === 'menuitem';
+		}
+		return withinSubset('button', role);
+	},
+	img: (node, role) => {
+		var hasAllowedEmptyAltRole = withinSubset('imgWithEmptyAlt', role);
+
+		if(node.alt) {
+      // any role except the roles used by empty alt values
+			return !hasAllowedEmptyAltRole;
+		} else {
+			return hasAllowedEmptyAltRole;
+		}
+	},
+	input: (node, role) => {
+		switch(node.type) {
+			case 'button':
+				return withinSubset('inputTypeButton', role);
+			case 'checkbox':
+				if(role === 'button' && node.hasAttribute('aria-pressed')) {
+					return true;
+				}
+				return withinSubset('inputTypeCheckbox', role);
+			case 'image':
+				return withinSubset('inputTypeImage', role);
+			case 'radio':
+				return role === 'menuitemradio';
+			default:
+				return false;
+		}
+	},
+	li: (node, role) => {
+		let hasImplicitListitemRole = node.matches('ol li, ul li');
+
+		if(hasImplicitListitemRole) {
+			return withinSubset('li', role);
+		} else {
+			return true;
+		}
+	},
+	link: (node) => !(node.href),
+	menu: (node) => {
+		if(node.type === 'context'){
+			return false;
+		}
+		return true;
+	},
+	option: (node) => {
+		let withinOptionList = node.matches('select > option, datalist > option, optgroup > option');
+		return !withinOptionList;
+	},
+	select: (node, role) => (!node.multiple && node.size <= 1 && role === 'menu'),
+	svg: (node, role) => {
+    // if in svg context it all roles may be used
+		if(node.parentNode && node.parentNode.namespaceURI === 'http://www.w3.org/2000/svg') {
+			return true;
+		}
+
+		return withinSubset('svg', role);
+	}
+};
+
+var roles = axe.commons.aria.getRoles(node);
+var implicitRole = axe.commons.aria.implicitRole(node);
+var tagName = axe.utils.getTagName(node);
+var disallowedRoles = [];
+
+// check if check on the element is disabled through the options
+if (options.disable && options.disable.indexOf(tagName) > -1) {
+	return true;
+}
+
+roles.forEach(function(role) {
+	var allowed = true;
+
+	// any implicit role is disallowed
+	if (!options.allowImplicit && role === implicitRole) {
+    // edge case: setting implicit role row on tr element is allowed when child of table[role='grid']
+		if (!(role === 'row' && tagName === 'tr' && node.matches('table[role="grid"] :scope'))) {
```

Comment:
An edge case where setting a implicit role is allowed, only applicable in the current spec when using `<table role='grid'><td role='cell'></td></table>`. I don't like the current location of the check...